src/back/wgsl/writer.rs

```diff
@@ -530,15 +671,79 @@ impl<W: Write> Writer<W> {
         }
 
         match *expression {
-            Expression::Constant(constant) => {
-                self.write_constant(&module.constants[constant], false)?
-            }
+            Expression::Constant(constant) => self.write_constant(module, constant)?,
             Expression::Compose { ty, ref components } => {
                 self.write_type(&module, ty)?;
                 write!(self.out, "(")?;
-                self.write_slice(components, |this, _, arg| {
-                    this.write_expr(&module, *arg, func_ctx)
-                })?;
+                // !spv-in specific notes!
+                // WGSL does not support all SPIR-V builtins and we should skip it in generated shaders.
+                // We already skip them when we generate struct type.
+                // Now we need to find components that used struct with ignored builtins.
+
+                // So, why we can't just return the error to a user?
+                // We can, but otherwise, we can't generate WGSL shader from any glslang SPIR-V shaders.
+                // glslang generates gl_PerVertex struct with gl_CullDistance, gl_ClipDistance and gl_PointSize builtin inside by default.
+                // All of them are not supported by WGSL.
+
+                // We need to copy components to another vec because we don't know which of them we should write.
+                let mut components_to_write = Vec::with_capacity(components.len());
+                for component in components {
+                    let mut skip_component = false;
+                    if let Expression::Load { pointer } = func_ctx.expressions[*component] {
+                        let pointer_global = &func_ctx.info[pointer].assignable_global;
```

Comment:
Any particular reason to use the `assignable_global` here?
Could we instead just match `function_ctx.expressions[pointer]` for `Expression::GlobalVariable` and get the handle from there?

Reply:
All required expressions here are `Expression::AccessIndex` and we need to find a global associated with this expression.

Reply:
do we expect these access chains to be more than 1 length?

Reply:
This handling of `Compose` looks really strange to me. I think we'd need to rewrite it entirely.
It's currently looking specifically for a global variable that is a struct in Private storage class with this member...
Here is what we can do instead:
 
Check the output type. If it's a struct, check the binding on the member of the struct. If the binding is an unsupported builtin, skip the field.

This logic should be much simpler and more general at the same time.